python_scripts/ensemble.py

```diff
@@ -0,0 +1,972 @@
+#   samples;
+# * one need to assign a weight to each learner when making the predictions.
+#
+# Here, we defined really simple scheme to define the weight. However, there is
+# some statistical basis which can be used to theoretically defined these
+# weights.
+
+# %%
+from sklearn.ensemble import AdaBoostClassifier
+
+adaboost = AdaBoostClassifier(
+    base_estimator=DecisionTreeClassifier(max_depth=3, random_state=0),
+    n_estimators=3,
+    algorithm="SAMME",
+    random_state=0,
+)
+adaboost.fit(X, y)
+
+_, axs = plt.subplots(ncols=3, figsize=(16, 6))
+
+for ax, tree in zip(axs, adaboost.estimators_):
+    plot_decision_function(X, y, tree, fit=False, ax=ax)
+
+print(f"Weight of each classifier: {adaboost.estimator_weights_}")
+print(f"Error of each classifier: {adaboost.estimator_errors_}")
+
+# %% [markdown]
+# We see that AdaBoost learnt 3 different classifiers focusing on different
+# samples. Looking at the weight of each learner, we see that we are more
+# confident in the first learner than the latest since the weights are
+# decreasing. It is in line of the error rate of each classifier.
+#
+# While AdaBoost is a nice algorithm to get the internal machinery of boosting
+# algorithms, this is not the most efficient machine-learning algorithm.
+# The most efficient algorithm based on boosting is the gradient-boosting
+# decision tree (GBDT) algorithm which we will present now.
+#
+# ### Gradient-boosting decision tree (GBDT)
+#
+# Gradient-boosting differs from AdaBoost since at each stage, it will fit
+# the residuals (hence the name "gradient") instead of the samples themselves.
+#
+# In this section, we will provide some intuitions regarding the way learners
+# will be combined to give the final prediction. In this regard, let's start
+# with our regression problem.
+
+# %%
+x, y = generate_data(sorted=True)
+
+plt.scatter(x, y, color="k", s=9)
+plt.xlabel("Feature")
+_ = plt.ylabel("Target")
+
+# %% [markdown]
+# As we previously discussed, boosting will be based on assembling a sequence
+# of learner. We will start by creating a decision tree regressor. We will fix
+# the depth of the tree such that the resulting learner will underfit the data.
+
+# %%
+tree = DecisionTreeRegressor(max_depth=3, random_state=0)
+tree.fit(x.reshape(-1, 1), y)
+
+grid = np.linspace(np.min(x), np.max(x), num=200)
+y_pred_grid_raw = tree.predict(grid.reshape(-1, 1))
+
+plt.scatter(x, y, color="k", s=9)
+plt.xlabel("Feature")
+plt.ylabel("Target")
+line_predictions = plt.plot(grid, y_pred_grid_raw, "--")
+
+y_pred_raw = tree.predict(x.reshape(-1, 1))
+for idx in range(len(y)):
+    lines_residuals = plt.plot(
+        [x[idx], x[idx]], [y[idx], y_pred_raw[idx]], color="red",
+    )
+
+_ = plt.legend(
+    [line_predictions[0], lines_residuals[0]], ["Fitted tree", "Residuals"]
+)
+
+# %% [markdown]
+# Since the tree is underfitting the data, the accuracy of the tree will is far
+# to be perfect on the training data. We can observe it on the figure by
```

Comment:
```suggestion
# Since the tree is underfitting the data, its accuracy will be far from
# perfect on the training data. We can observe it on the figure by
```